### PR TITLE
Improve performance of SyntaxTreeIndex checksum computation.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This drops the checksum computation portion NavTo from around 14% of the scenario to around 8%.  It's still pretty expensive as we have to still read in the entire file and checksum it to know if we can actually use our cached index.  However, we at least don't need to also compute checksums for things like metadata/analyzers which cannot affect us.

This helps after first load when we have to load all hte indices to search them.

### CODE REVIEW

File: src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs [hunk 1]
```diff
@@ -54,8 +54,12 @@ internal sealed partial class SyntaxTreeIndex : IObjectWritable
...
             // We also want the checksum to change any time our serialization format changes.  If
             // the format has changed, all previous versions should be invalidated.
-            var projectChecksumState = await document.Project.State.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
-            var parseOptionsChecksum = projectChecksumState.ParseOptions;
+            var project = document.Project;
+
+            // This is the same logic the project itself uses here:
+            // https://github.com/dotnet/roslyn/blob/5e725e29186cc1b341b1a9b2baae73e34d245232/src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs#L48
+            var serializer = project.Solution.Workspace.Services.GetService<ISerializerService>();
+            var parseOptionsChecksum = ChecksumCache.GetOrCreate(project.ParseOptions, o => serializer.CreateChecksum(o, cancellationToken));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs:62]: the previous approach was simple and clearer.  however, it was about doubly expensive as it also required computed all the *other* checksums for the project, even though we only ever accessed the ParseOptions portion of it.

File: src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs [hunk 2]
```diff
@@ -54,8 +54,12 @@ internal sealed partial class SyntaxTreeIndex : IObjectWritable
...
             // We also want the checksum to change any time our serialization format changes.  If
             // the format has changed, all previous versions should be invalidated.
-            var projectChecksumState = await document.Project.State.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
-            var parseOptionsChecksum = projectChecksumState.ParseOptions;
+            var project = document.Project;
+
+            // This is the same logic the project itself uses here:
+            // https://github.com/dotnet/roslyn/blob/5e725e29186cc1b341b1a9b2baae73e34d245232/src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs#L48
+            var serializer = project.Solution.Workspace.Services.GetService<ISerializerService>();
+            var parseOptionsChecksum = project.SupportsCompilation ? ChecksumCache.GetOrCreate(project.ParseOptions, _ => serializer.CreateChecksum(project.ParseOptions, cancellationToken)) : Checksum.Null;
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs:62]: ❔ Can we make this a static method on `ProjectState`?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs]: Done!

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs [hunk 3]
```diff
@@ -27,6 +27,14 @@ public async Task<Checksum> GetChecksumAsync(CancellationToken cancellationToken
             return collection.Checksum;
         }
 
+        public Checksum GetParseOptionsChecksum(CancellationToken cancellationToken)
+            => GetParseOptionsChecksum(_solutionServices.Workspace.Services.GetService<ISerializerService>(), cancellationToken);
```
Comment (dibarbet) [inline src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs:31]: could this be nullable enabled, and then GetRequiredService?

File: src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs (hunk 2 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex_Persistence.cs]: Yeah that was a big improvement!

File: src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs:31]: i tried, but the impact was very large on this file.  and i don't like doing NRT on a file in piecemeal fashion.  Thanks!

**Pull request merged**